local-volume/provisioner/pkg/deleter/jobcontroller.go

```diff
@@ -0,0 +1,354 @@
+
+	// runWorker will loop until "something bad" happens.  The .Until will
+	// then rekick the worker after one second
+	wait.Until(c.runWorker, time.Second, stopCh)
+}
+
+func (c *jobController) runWorker() {
+	for c.processNextItem() {
+	}
+}
+
+// processNextWorkItem serially handles the events provided by the informer.
+func (c *jobController) processNextItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+
+	defer c.queue.Done(key)
+
+	err := c.processItem(key.(string))
+	if err == nil {
+		// No error, tell the queue to stop tracking history
+		c.queue.Forget(key)
+	} else if c.queue.NumRequeues(key) < maxRetries {
+		glog.Errorf("Error processing %s (will retry): %v", key, err)
+		// requeue the item to work on later
+		c.queue.AddRateLimited(key)
+	} else {
+		// err != nil and too many retries
+		glog.Errorf("Error processing %s (giving up): %v", key, err)
+		c.queue.Forget(key)
+		utilruntime.HandleError(err)
+	}
+
+	return true
+}
+
+func (c *jobController) processItem(key string) error {
+	glog.Infof("Processing change to Pod %s", key)
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	job, err := c.jobLister.Jobs(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		glog.Infof("Job %s has been deleted", key)
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("Error fetching object with key %s from store: %v", key, err)
+	}
+
+	if job.DeletionTimestamp != nil {
+		// This is just an event in response to the deletion of the job. Job deletion is probably still in progress.
+		glog.Infof("Job %s deletion timestamp has been set (%s)", key,
+			*job.DeletionTimestamp)
+		return nil
+	}
+
+	if job.Status.Succeeded == 0 {
+		glog.Infof("Job %s has not yet completed successfully", key)
+	} else {
+		pvName := getPVNameFromJob(job.Name)
+		// Order here is important. Delete the finished job first and then the PV.
+		glog.Infof("Job %s has completed, deleting the job", key)
+		if err := c.RuntimeConfig.APIUtil.DeleteJob(job.Name, c.namespace); err != nil {
+			if !errors.IsNotFound(err) {
+				return fmt.Errorf("Error deleting Job %q: %v", job.Name, err.Error())
+			}
+			// If the job has been deleted, we assume that the PV has also been recycled by the same processItem()
+			// call that deleted the job. There is a chance that the thread failed to delete the PV after deleting
+			// the job. In that case, a new cleaning job will eventually be launched and the whole cleaning cycle
+			// will be repeated. This is an acceptable inefficiency for this failure case, as making sure that
+			// the deletion of job and pv is done by the same processItem() call and in that fixed sequence,
+			// avoids a whole bunch of unwanted race conditions.
+			glog.Infof("Job %s has already been deleted.", key)
+			return nil
+		}
+
+		glog.Infof("Job %s has completed and been deleted. Deleting PV %s", key, pvName)
+		if err := c.APIUtil.DeletePV(pvName); err != nil {
```

Comment:
Instead of deleting the PV in the job controller, can you signal that the job is done in the proc table, and have the deleter delete the API object?  Then the PV object deletion is handled the same regardless of if you use jobs or not.

Reply:
Fixed.